core/src/main/scala/kafka/server/Kip500Broker.scala

```diff
@@ -17,16 +17,77 @@
 
 package kafka.server
 
+import java.util.concurrent.locks.ReentrantLock
+
 import kafka.metrics.KafkaMetricsReporter
 import org.apache.kafka.common.utils.Time
+import org.apache.kafka.metadata
 
 /**
  * A KIP-500 Kafka broker.
  */
 class Kip500Broker(config: KafkaConfig,
                    time: Time,
                    threadNamePrefix: Option[String],
-                   kafkaMetricsReporters: Seq[KafkaMetricsReporter])
-                     extends KafkaBroker(config, time, threadNamePrefix,
-                       kafkaMetricsReporters) {
+                   kafkaMetricsReporters: Seq[KafkaMetricsReporter]) extends KafkaBroker {
+  import kafka.server.KafkaServerManager._
+
+  val lock = new ReentrantLock()
+  val awaitShutdownCond = lock.newCondition()
+  var status: ProcessStatus = KafkaServerManager.SHUTDOWN
+
+  private def maybeChangeStatus(from: ProcessStatus, to: ProcessStatus): Boolean = {
+    lock.lock()
+    try {
+      if (status != from) return false
+      status = to
+      if (to == SHUTDOWN) awaitShutdownCond.signalAll()
+    } finally {
+      lock.unlock()
+    }
+    true
+  }
+
+  override def startup(): Unit = {
+    if (!maybeChangeStatus(SHUTDOWN, STARTING)) return
+    try {
+      maybeChangeStatus(STARTING, STARTED)
+    } catch {
+      case e: Throwable =>
+        maybeChangeStatus(STARTING, STARTED)
+        fatal("Fatal error during controller startup. Prepare to shutdown", e)
+        shutdown()
+        throw e
+    }
+  }
+
+  override def shutdown(): Unit = {
+    if (!maybeChangeStatus(STARTED, SHUTTING_DOWN)) return
+    try {
+    } catch {
+      case e: Throwable =>
+        fatal("Fatal error during broker shutdown.", e)
+        throw e
+    } finally {
+      maybeChangeStatus(SHUTTING_DOWN, SHUTDOWN)
+    }
+  }
+
+  override def awaitShutdown(): Unit = {
+    lock.lock()
+    try {
+      while (true) {
+        if (status == SHUTDOWN) return
+        awaitShutdownCond.awaitUninterruptibly()
+      }
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  override def metrics() = null
+
+  override def currentState(): metadata.BrokerState = null
```

Comment:
In combination with import change above:

```suggestion
  override def currentState(): BrokerState = null
```